lib/api/axiosFactory.js

```diff
@@ -0,0 +1,274 @@
+const { firmCredentials } = require("./firmCredentials");
+const axios = require("axios");
+const { consola } = require("consola");
+const pkg = require("../../package.json");
+
+class AxiosFactory {
+  constructor() {}
+
+  /**
+   * Create an axios instance for a given type and environment id.
+   * It will add the necessary headers and interceptors to handle the authorization, and refresh the tokens if needed.
+   * The created instance will be used to make requests to the Silverfin API.
+   * @param {String} type - The type of instance to create (firm or partner)
+   * @param {Number} envId - The environment id to create the instance for
+   * @returns {Object} - The created axios instance
+   */
+  static createInstance(type, envId) {
+    this.BASE_URL = firmCredentials.getHost();
+    let axiosInstance;
+
+    switch (type) {
+      case "firm":
+        axiosInstance = this.#createAxiosInstanceForFirms(envId);
+        break;
+      case "partner":
+        axiosInstance = this.#createAxiosInstanceForPartners(envId);
+        break;
+      default:
+        consola.error(`Invalid type environment: ${type}`);
+        process.exit(1);
+    }
+
+    return axiosInstance;
+  }
+
+  // PRIVATE METHODS
+
+  static #createAxiosInstanceForFirms(envId) {
+    const firmTokens = firmCredentials.getTokenPair(envId);
+    if (!firmTokens) {
+      consola.error(`Missing authorization for firm id: ${envId}`);
+      process.exit(1);
+    }
+
+    let axiosDetails;
+    if (this.#isStaging()) {
+      axiosDetails = {
+        baseURL: `${this.BASE_URL}/api/v4/f/${envId}`,
+        headers: {
+          "User-Agent": `silverfin-cli/${pkg.version}`,
+          Authorization: this.#basicAuthHeader(),
+        },
+        params: {
+          access_token: firmTokens.accessToken,
+        },
+      };
+    } else {
+      axiosDetails = {
+        baseURL: `${this.BASE_URL}/api/v4/f/${envId}`,
+        headers: {
+          "User-Agent": `silverfin-cli/${pkg.version}`,
+          Authorization: `Bearer ${firmTokens.accessToken}`,
```

Comment:
Why not send accessToken as a `param` for both staging and production? It'd make it more obvious that the only difference between staging and production is really basic auth. Otherwise the regular API authentication works exactly the same.

Reply:
> Isn't headers preferred over param?

Haven't read anything suggesting that. Do you have any reference?

Reply:
I'd guess it's more of a convention than any real benefit, but let's leave it as is then.